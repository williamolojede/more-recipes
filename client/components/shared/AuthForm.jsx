import React from 'react';
import PropTypes from 'prop-types';

/**
 * @class AuthForm
 * @extends {Component}
 */
class AuthForm extends React.Component {
  /**
   * @param {node} e - element
   * @return {view} - renders a form
   */
  onFormSubmit(e) {
    e.preventDefault();

    const email = this.emailInput.value;
    const password = this.passwordInput.value;
    if (this.props.type === 'login') {
      this.props.authFormSubmit({ email, password });
    }

    this.authForm.reset();
  }

  /**
   * @return {view} - renders a form
   */
  render() {
    return (
      <form
        className="auth-form form"
        ref={(node) => { this.authForm = node; }}
        onSubmit={e => this.onFormSubmit(e)}
      >
        {
          this.props.type === 'signup' ?
            <input
              type="text"
              name="usr-name"
              className="form__input"
              placeholder="name"
              required
            />
            :
            null
        }
        <input
          id="usr-email"
          type="email"
          name="usr-email"
          className="form__input"
          placeholder="email"
          required
          ref={(node) => { this.emailInput = node; }}
        />
        <input
          id="usr-pswd"
          type="password"
          name="usr-pswd"
          className="form__input"
          placeholder="password"
          required
          ref={(node) => { this.passwordInput = node; }}
        />

        <button type="submit" className="form__submit z-depth-1">
          {
            this.props.type === 'signup' ?
              'Register'
              :
              'Login'
          }
        </button>
      </form>
    );
  }
}

AuthForm.propTypes = {
  type: PropTypes.string.isRequired,
  authFormSubmit: PropTypes.func.isRequired,
};

export default AuthForm;
