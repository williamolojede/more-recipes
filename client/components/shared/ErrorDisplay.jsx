import React from 'react';
import { errorDisplayPropTypes } from '../../config/proptypes';

const ErrorDisplay = props => (
  <div className="error z-depth-1">
    <span className="error__cancel">X</span>
    <p className="error__message">{ props.message }</p>
  </div>
);

ErrorDisplay.propTypes = errorDisplayPropTypes;

export default ErrorDisplay;
