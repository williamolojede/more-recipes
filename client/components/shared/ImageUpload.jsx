import React, { Component } from 'react';
import PropTypes from 'prop-types';
import ReactLoading from 'react-loading';

class ImageUpload extends Component {
  render() {
    const {
      imgUrl,
      altText,
      imageUploadProgress,
      className,
      defaultImgUrl,
      imageUploadHandler
    } = this.props;

    return (
      <div className={`image-upload ${className}`}>
        <div className="image-upload__preview">
          <img
            src={imgUrl || defaultImgUrl}
            alt={altText}
            className="image-upload__preview___image"
          />
          <div className="image-upload__progress">
            {
              imageUploadProgress === 'started' &&
              <ReactLoading
                type="spokes"
                color="#000"
                height="65px"
                width="65px"
                className="loading-icon"
              />
            }
          </div>
        </div>
        <label htmlFor="recipe-image" className="z-depth-1">Upload new picture</label>
        <input
          type="file"
          id="recipe-image"
          name="filename"
          accept="image/gif, image/jpeg, image/png"
          ref={node => this.recipeImageInput = node}
          onChange={() => imageUploadHandler(this.recipeImageInput)}
        />
      </div>
    );
  }
}

ImageUpload.defaultProps = {
  className: '',
  imgUrl: ''
};

ImageUpload.propTypes = {
  className: PropTypes.string,
  imgUrl: PropTypes.string,
  altText: PropTypes.string.isRequired,
  defaultImgUrl: PropTypes.string.isRequired,
  imageUploadProgress: PropTypes.oneOf(['pending', 'started', 'failed', 'completed']).isRequired,
  imageUploadHandler: PropTypes.func.isRequired
};

export default ImageUpload;
