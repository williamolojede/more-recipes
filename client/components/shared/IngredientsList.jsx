import React from 'react';

const IngredientsList = () => (
  <div className="recipe__ingredients col s12 m5">
    <h3>Ingredients</h3>
    <ul>
      <li><span>1 1/2 cups</span> all-purpose flour</li>
      <li><span>1 1/4 cups</span> milk</li>
      <li><span>3 1/2 teaspoons</span> baking powder</li>
      <li><span>1</span> egg</li>
      <li><span>1 teaspoon</span> salt</li>
      <li><span>3 tablespoons</span> butter, melted</li>
      <li><span>1 tablespoon </span> white sugar</li>
    </ul>
  </div>
);

export default IngredientsList;
