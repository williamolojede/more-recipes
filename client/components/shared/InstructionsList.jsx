import React from 'react';

const InstructionsList = () => (
  <div className="recipe__instructions col s12 m6">
    <h3>Instructions</h3>
    <ol>
      <li>
        <span>In a large bowl, sift together the flour, baking powder, salt and sugar. Make a well in the center and pour in the milk, egg and melted butter; mix until smooth.</span>
      </li>
      <li>
        Heat a lightly oiled griddle or frying pan over medium high heat. Pour or scoop the batter onto the griddle, using approximately 1/4 cup for each pancake. Brown on both sides and serve hot.
      </li>
    </ol>
  </div>
);

export default InstructionsList;
