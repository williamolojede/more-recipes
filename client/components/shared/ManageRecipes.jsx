import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { NavLink } from 'react-router-dom';

import ConnectedSiteNav from '../shared/SiteNav';
import SiteFooter from '../shared/SiteFooter.jsx';
import Notification from '../shared/Notification.jsx';

import { currentUserPropTypes } from '../../config/proptypes';

class ManageRecipes extends Component {
  render() {
    return (
      <div className="page page-manage">
        <header className="site-header">
          <ConnectedSiteNav currentUser={this.props.currentUser} />
        </header>
        <main>
          <div className="page-manage__header-image" />
          <ul className="page-manage__nav tabs z-depth-1">
            <li className="tab col s4">
              <NavLink exact to="/user/recipes/create" activeClassName="active">Add</NavLink>
            </li>
            <li className="tab col s4">
              <NavLink exact to="/user/recipes" activeClassName="active">Recipes</NavLink>
            </li>
            <li className="tab col s4">
              <NavLink exact to="/user/favorites" activeClassName="active">Favorites</NavLink>
            </li>
          </ul>
          {
            this.props.children
          }
        </main>
        <Notification notification={this.props.notification} dispatch={this.props.dispatch} />
        <SiteFooter />
      </div>
    );
  }
}
ManageRecipes.propTypes = {
  dispatch: PropTypes.func.isRequired,
  ...currentUserPropTypes,
  children: PropTypes.element.isRequired,
  notification: PropTypes.string.isRequired
};

export const mapStateToProps = ({
  auth: { currentUser },
  notification
}) => ({
  currentUser,
  notification
});

export { ManageRecipes as PureManageRecipes };
export default connect(mapStateToProps)(ManageRecipes);
