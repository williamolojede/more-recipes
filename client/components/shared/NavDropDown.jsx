import React from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';

const NavDropDown = ({ fullname, handleLogout }) => (
  <ul className="dropdown-menu z-depth-1">
    <li className="dropdown-header">Signed in as
      <p>{ fullname }</p>
    </li>
    <li className="dropdown-item divider" />
    <li className="dropdown-item"><Link to="/user">Your Profile</Link></li>
    <li className="dropdown-item"><Link to="/user/recipes">Your Recipes</Link></li>
    <li className="dropdown-item"><Link to="/user/favorites">Your Favorite Recipes</Link></li>
    <li className="dropdown-item divider" />
    <li className="dropdown-item"><Link to="/user/recipes/create">Create New Recipe</Link></li>
    <li className="dropdown-footer">
      <Link
        to="/logout"
        className="logout__button"
        onClick={handleLogout}
      >
        Sign Out
      </Link>
    </li>
  </ul>
);

NavDropDown.propTypes = {
  fullname: PropTypes.string.isRequired,
  handleLogout: PropTypes.func.isRequired
};

export default NavDropDown;
