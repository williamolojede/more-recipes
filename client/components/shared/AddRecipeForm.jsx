import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';

import defaultConfig from '../../config/default';
import { recipePropTypes } from '../../config/proptypes';
import {
  addRecipe,
  updateRecipe,
  fetchSingleRecipe,
  resetSingleRecipe
} from '../../actions/recipe';
import imageUpload from '../../utils/imageUpload';

import ErrorDisplay from './ErrorDisplay';
import Preloader from './Preloader';
import ImageUpload from './ImageUpload';
import SubmitButton from './SubmitButton';


class AddRecipeForm extends Component {
  constructor(props) {
    super(props);

    this.state = {
      recipe: this.props.recipe,
      renderInput: {
        ingredient: false,
        instruction: false,
      },
      submitLoading: false,
      // pending, started, failed, completed
      imageUploadProgress: 'pending',
      addRecipe: {
        allow: true,
        errorMessage: '',
      }
    };
  }

  componentDidMount() {
    if (this.props.isEditMode) {
      this.props.dispatch(fetchSingleRecipe(this.props.match.params.recipeId));
    }
  }

  componentWillReceiveProps(newProps) {
    this.setState({ recipe: newProps.recipe });
  }

  componentWillUnmount() {
    this.props.dispatch(resetSingleRecipe());
  }

  /**
   * @param {object} event  - event object
   * @param {object} el - input element
   * @param {string} type - recipe key name whose value is to be updated
   * @return {undefined}
   */
  addItem = (event, el, type) => {
    if (event.keyCode === 13) {
      event.preventDefault();
      const recipe = Object.assign({}, this.state.recipe);
      const text = el.value.trim();
      if (text) {
        recipe[type] = recipe[type].concat(text);
        this.setState({ recipe });
      }
      el.value = '';
    }
  }

  /**
   * @param {object} event  - event object
   * @param {string} type - recipe key name whose value is to be updated
   * @param {number} index - input element
   * @return {undefined}
   */
  removeItem = (event, type, index) => {
    event.preventDefault();
    const recipe = Object.assign({}, this.state.recipe);

    recipe[type] = [
      ...recipe[type].slice(0, index),
      ...recipe[type].slice(index + 1)
    ];
    this.setState({ recipe });
  }

  /**
   * Form submission handler
   * @return {undefined}
   */
  addRecipe = () => {
    const {
      name,
      description,
      ingredients,
      instructions
    } = this.state.recipe;

    if (!name || !description) {
      this.setState({
        addRecipe: {
          allow: false,
          errorMessage: 'recipe name and description are required'
        }
      });
    } else if (!ingredients.length || !instructions.length) {
      this.setState({
        addRecipe: {
          allow: false,
          errorMessage: 'ingredients and instructions cannot be empty'
        }
      });
    } else {
      this.setState({ submitLoading: true });
      this.props.dispatch(addRecipe(this.state.recipe))
        .then(this.handleFormSubmitSuccess);
    }
  }

  updateRecipe = () => {
    this.setState({ submitLoading: true });
    this.props.dispatch(updateRecipe(this.state.recipe))
      .then(this.handleFormSubmitSuccess);
  }

  handleFormSubmitSuccess = () => {
    // set submit loading to false and redirect to
    //  all recipes page after .5s
    this.setState({ submitLoading: false });
    setTimeout(() => this.props.history.push('/user/recipes'), 500);
  }

  submitForm = (e) => {
    e.preventDefault();
    if (this.props.isEditMode) {
      this.updateRecipe();
    } else {
      this.addRecipe();
    }
  }

  /**
   * @param {object} e - event object
   * @param {string} type - type of input to be show, for ingredients or instructions
   * @return {undefined}
   */
  showInput = (e, type) => {
    e.preventDefault();
    const renderInput = Object.assign({}, this.state.renderInput);

    // if already true do nothing
    if (!renderInput[type]) {
      renderInput[type] = true;
      this.setState({ renderInput });
    }
  }

  /**
   * @param {object} e - event object
   * @param {string} type - type of input to be show, for ingredients or instructions
   * @return {undefined}
   */
  hideInput = (e, type) => {
    const renderInput = Object.assign({}, this.state.renderInput);
    // if already true do nothing
    if (renderInput[type]) {
      renderInput[type] = false;
      this.setState({ renderInput });
    }
  }

  /**
   * Handles image upload
   * @param {object} imageInput - the image input to get image to be uploaded from
   * @return {undefined}
   */
  imageUploadHandler = (imageInput) => {
    imageUpload.call(
      this,
      imageInput,
      `recipes/${Date.now()}`,
      (downloadURL) => {
        const recipe = Object.assign({}, this.state.recipe);
        recipe.img_url = downloadURL;
        this.setState({ recipe, imageUploadProgress: 'completed' });
      }
    );
  }

  /**
   * Adds the recipe name or description to the state
   * @param {object} el - element
   * @param {string} type - property name whose value is to be added, type: 'name' or 'description'
   * @return {undefined}
   */
  addRecipeProp = (el, type) => {
    const text = el.value.trim();
    const recipe = Object.assign({}, this.state.recipe);
    recipe[type] = text;
    this.setState({ recipe });
  }

  /**
   * renderItems returns of a list of item that matches the provide type in state.recipe
   * @param {string} type - the type of item in state to be looped through
   * @return {jsx} - a list of items to be redered
   */
  renderItems = type => this.state.recipe[type].map((item, index) => (
    <li key={index}>
      {item}
      <a href="/delete" className={`remove-item remove-${type}`} onClick={e => this.removeItem(e, type, index)}>
        <i className="mdi mdi-delete" />
      </a>
    </li>
  ))

  renderForm = () => (
    <form
      className="my-recipe__add--form"
      onSubmit={this.submitForm}
    >
      <label className="my-recipe__add--form___name" htmlFor="recipe-name">
          Name
        <input
          type="text"
          id="recipe-name"
          name="recipe-name"
          placeholder={this.state.recipe.name}
          defaultValue={this.state.recipe.name}
          ref={node => this.recipeNameInput = node}
          onChange={() => this.addRecipeProp(this.recipeNameInput, 'name')}
        />
      </label>
      {/*  description-image wrapper  */}
      <div className="description-image">
        <label className="my-recipe__add--form___desc" htmlFor="recipe-description">
            Description
          <textarea
            placeholder="e.g: short description..."
            id="recipe-description"
            defaultValue={this.state.recipe.description}
            ref={node => this.recipeDescInput = node}
            onChange={() => this.addRecipeProp(this.recipeDescInput, 'description')}
          />
        </label>
        <div className="my-recipe__add--form___img">
          <p className="label">Image</p>
          <ImageUpload
            imgUrl={this.state.recipe.img_url}
            altText={this.state.recipe.name}
            imageUploadProgress={this.state.imageUploadProgress}
            imageUploadHandler={this.imageUploadHandler}
            defaultImgUrl={defaultConfig.recipeImgUrl}
          />
        </div>
      </div>

      {/* ingredients-instruction wrapper  */}
      <div className="ingredients-instruction">
        <fieldset className="ingredients-form my-recipe__add--form___ingredients">
          <legend>Ingredient(s):</legend>
          <ol>
            {
              this.renderItems('ingredients')
            }
          </ol>
          {
            this.state.renderInput.ingredient ?
              <textarea
                type="text"
                name="recipe-ingredient__name"
                className="ingredient-instrution-input"
                placeholder="e.g: egusi"
                ref={node => this.addIngredientInput = node}
                onKeyUp={e => this.addItem(e, this.addIngredientInput, 'ingredients')}
                onBlur={e => this.hideInput(e, 'ingredient')}
              />
              :
              <a href="/add-new" className="add-new" onClick={e => this.showInput(e, 'ingredient')}>
                <i className="mdi mdi-plus" />
                <span>Add an ingredient</span>
              </a>
          }
        </fieldset>
        <fieldset className="instructions-form my-recipe__add--form___instructions">
          <legend>Instruction(s):</legend>
          <ol>
            {
              this.renderItems('instructions')
            }
          </ol>
          {
            this.state.renderInput.instruction ?
              <textarea
                type="text"
                name="recipe-instruction__name"
                className="ingredient-instrution-input"
                placeholder="e.g: boil water"
                ref={node => this.addInstructionInput = node}
                onKeyUp={e => this.addItem(e, this.addInstructionInput, 'instructions')}
                onBlur={e => this.hideInput(e, 'instruction')}
              />
              :
              <a href="/add-new" className="add-new" onClick={e => this.showInput(e, 'instruction')}>
                <i className="mdi mdi-plus" />
                <span>Add an instruction</span>
              </a>
          }
        </fieldset>
      </div>
      <div className="form-btns">
        <SubmitButton
          submitText={this.props.isEditMode ? 'Update' : 'Submit'}
          submitLoading={this.state.submitLoading}
        />
        {
          this.props.isEditMode ?
            <Link to="/user/recipes">or go back to your recipe list</Link>
            :
            null
        }
      </div>
    </form>
  )

  render() {
    return (
      <div className="container page-manage__add z-depth-1">
        {
          !this.state.addRecipe.allow &&
            <ErrorDisplay message={this.state.addRecipe.errorMessage} />
        }
        {
          (this.props.isEditMode && this.props.isFetching) ? <Preloader />
            : this.renderForm()
        }
      </div>
    );
  }
}

AddRecipeForm.propTypes = {
  dispatch: PropTypes.func.isRequired,
  ...recipePropTypes,
  isEditMode: PropTypes.bool.isRequired,
};

const mapStateToProps = ({
  isFetching,
  singleRecipe: { recipe }
}) => ({
  isFetching,
  recipe
});

export { AddRecipeForm as PureAddRecipeForm };
export default connect(mapStateToProps)(AddRecipeForm);
