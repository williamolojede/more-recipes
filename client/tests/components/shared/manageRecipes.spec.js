import React from 'react';
import { PureManageRecipes, mapStateToProps } from '../../../components/shared/ManageRecipes';
import userMock from '../../__mocks__/user.mock';
import notifiacationMock from '../../__mocks__/notification.mock';

function setup() {
  const props = {
    dispatch: jest.fn(),
    currentUser: userMock,
    children: (() => (<h1>Yo</h1>))(),
    notification: notifiacationMock
  };
  const wrapper = shallow(<PureManageRecipes {...props} />);

  return {
    wrapper
  };
}

describe('Manage Recipes', () => {
  it('should render with right amount of elements', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  });

  describe('mapStateToProps', () => {
    it('should return an object with props required by component from store', () => {
      const store = {
        auth: {
          currentUser: userMock,
        },
        notification: 'recipe created sucesfully'
      };
      expect(mapStateToProps(store).notification).toBe('recipe created sucesfully');
      expect(mapStateToProps(store).currentUser.fullname).toBe(userMock.fullname);
    });
  });
});
