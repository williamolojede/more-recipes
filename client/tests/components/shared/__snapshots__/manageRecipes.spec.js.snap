// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Manage Recipes should render with right amount of elements 1`] = `
<div
  className="page page-manage"
>
  <header
    className="site-header"
  >
    <Connect(SiteNav)
      currentUser={
        Object {
          "fullname": "John Doe",
          "id": 1,
        }
      }
    />
  </header>
  <main>
    <div
      className="page-manage__header-image"
    />
    <ul
      className="page-manage__nav tabs z-depth-1"
    >
      <li
        className="tab col s4"
      >
        <NavLink
          activeClassName="active"
          ariaCurrent="true"
          exact={true}
          to="/user/recipes/create"
        >
          Add
        </NavLink>
      </li>
      <li
        className="tab col s4"
      >
        <NavLink
          activeClassName="active"
          ariaCurrent="true"
          exact={true}
          to="/user/recipes"
        >
          Recipes
        </NavLink>
      </li>
      <li
        className="tab col s4"
      >
        <NavLink
          activeClassName="active"
          ariaCurrent="true"
          exact={true}
          to="/user/favorites"
        >
          Favorites
        </NavLink>
      </li>
    </ul>
    <h1>
      Yo
    </h1>
  </main>
  <Notification
    dispatch={[Function]}
    notification={
      Object {
        "message": "",
        "type": "success",
      }
    }
  />
  <SiteFooter />
</div>
`;
